core/Piwik.php

```diff
@@ -563,7 +563,14 @@ public static function redirectToModule($newModule, $newAction = '', $parameters
      */
     public static function isValidEmailString($emailAddress)
     {
-        return (preg_match('/^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9_.-]+\.[a-zA-Z]{2,}$/D', $emailAddress) > 0);
+        static $zendEmailValidator;
+        if(!$zendEmailValidator) {
+            $zendEmailValidator = new \Zend_Validate_EmailAddress(array(
+                'mx' => false,
+                'deep' => false
+            ));
+        }
```

Comment:
Most of what's in DI is currently not API, and when things are made API they are explicitly mentioned in docs (or should be).

As for the flyweight pattern, that normally involves more than one object. This is caching a single instance for the current request. If you don't want to use DI, you could use the transient cache, via, `Piwik\Cache::getTransientCache()`.

Whatever you do, please make sure this value can be reset during tests, and if possible, overridden (though I think this is only do-able through DI). Static local variables have made me waste days debugging tests, and I would gladly anger the design pattern gods or whatever entity is responsible for programming zealotry if it meant I didn't have to waste my time on pointless work.